Tries to get targets list after cmake invocation has failed

I haven't got a MWE for this, but we're seeing an issue sometimes where the CMake invocation fails, but the error message is unhelpful because it pertains to `make help` having failed. I believe this logic here should check the value of `rc` before calling `has_target`, or else the inner `get_makefile_targets` function should catch the exception and return an empty list in that case:

https://github.com/colcon/colcon-ros/blob/d77a37bcf0dbd6de92f5ce9cdb21efcb50dfd054/colcon_ros/task/catkin/build.py#L75-L82

Example:

```
-- Configuring incomplete, errors occurred!
See also "[...]/ws/build/python_orocos_kdl/CMakeFiles/CMakeOutput.log".
See also "[...]/ws/build/python_orocos_kdl/CMakeFiles/CMakeError.log".
Traceback (most recent call last):
  File "[...]/colcon-env/lib/python3.8/site-packages/colcon_core/executor/__init__.py", line 91, in __call__
    rc = await self.task(*args, **kwargs)
  File "[...]/colcon-env/lib/python3.8/site-packages/colcon_core/task/__init__.py", line 93, in __call__
    return await task_method(*args, **kwargs)
  File "[...]/colcon-env/lib/python3.8/site-packages/colcon_ros/task/catkin/build.py", line 82, in build
    if await has_target(args.build_base, 'install'):
  File "[...]/colcon-env/lib/python3.8/site-packages/colcon_cmake/task/cmake/__init__.py", line 57, in has_target
    return target in await get_makefile_targets(path)
  File "[...]/colcon-env/lib/python3.8/site-packages/colcon_cmake/task/cmake/__init__.py", line 77, in get_makefile_targets
    output = await check_output([
  File "[...]/colcon-env/lib/python3.8/site-packages/colcon_core/subprocess.py", line 109, in check_output
    assert not rc, \
AssertionError: Expected ['/usr/bin/cmake', '--build', '[...]/ws/build/python_orocos_kdl', '--target', 'help'] to pass: make: *** No rule to make target 'help'.  Stop.
```

FYI @iwanders

--- a/colcon_ros/task/ament_cmake/build.py
+++ b/colcon_ros/task/ament_cmake/build.py
@@ -60,10 +60,18 @@ class AmentCmakeBuildTask(TaskExtensionPoint):
             skip_hook_creation=False,
             environment_callback=add_app_to_cpp)
 
+        # if the build has failed getting targets might not be possible
+        try:
+            has_install_target = await has_target(args.build_base, 'install')
+        except Exception:
+            if not rc:
+                raise
+            has_install_target = False
+
         # add a hook for each available shell
         # only if the package has an install target
         additional_hooks = []
-        if await has_target(args.build_base, 'install'):
+        if has_install_target:
             shell_extensions = get_shell_extensions()
             file_extensions = []
             for shell_extensions_same_prio in shell_extensions.values():

--- a/colcon_ros/task/catkin/build.py
+++ b/colcon_ros/task/catkin/build.py
@@ -75,11 +75,19 @@ class CatkinBuildTask(TaskExtensionPoint):
         rc = await extension.build(
             skip_hook_creation=True, additional_targets=additional_targets)
 
+        # if the build has failed getting targets might not be possible
+        try:
+            has_install_target = await has_target(args.build_base, 'install')
+        except Exception:
+            if not rc:
+                raise
+            has_install_target = False
+
         # for catkin packages add additional hooks after the package has
         # been built and installed depending on the installed files
         # only if the package has an install target
         additional_hooks = []
-        if await has_target(args.build_base, 'install'):
+        if has_install_target:
             additional_hooks += create_environment_hook(
                 'ros_package_path', Path(args.install_base),
                 self.context.pkg.name, 'ROS_PACKAGE_PATH', 'share',

--- a/colcon_ros/task/cmake/build.py
+++ b/colcon_ros/task/cmake/build.py
@@ -35,8 +35,16 @@ class CmakeBuildTask(TaskExtensionPoint):
         rc = await extension.build(
             skip_hook_creation=True, environment_callback=add_app_to_cpp)
 
+        # if the build has failed getting targets might not be possible
+        try:
+            has_install_target = await has_target(args.build_base, 'install')
+        except Exception:
+            if not rc:
+                raise
+            has_install_target = False
+
         additional_hooks = []
-        if await has_target(args.build_base, 'install'):
+        if has_install_target:
             additional_hooks += create_pkg_config_path_environment_hooks(
                 Path(args.install_base), self.context.pkg.name)
 

